cmd/snap-confine/udev-support.c

```diff
@@ -37,6 +38,140 @@
 __attribute__((format(printf, 2, 3)))
 static void sc_dprintf(int fd, const char *format, ...);
 
+typedef struct sc_cgroup_fds {
+	int devices_allow_fd;
+	int devices_deny_fd;
+	int cgroup_procs_fd;
+} sc_cgroup_fds;
+
+static sc_cgroup_fds sc_cgroup_fds_new(void)
+{
+	sc_cgroup_fds empty = { -1, -1, -1 };
+	return empty;
+}
+
+typedef struct sc_udev_device_group {
+	bool is_v2;
+	char *security_tag;
+	union {
+		struct {
+			sc_cgroup_fds fds;
+		} v1;
+	};
+} sc_device_cgroup;
+
+static sc_cgroup_fds sc_udev_open_cgroup_v1(const char *security_tag);
+static void sc_cleanup_cgroup_fds(sc_cgroup_fds * fds);
+
+static void _sc_cgroup_v1_init(sc_device_cgroup * self)
+{
+	self->v1.fds = sc_cgroup_fds_new();
+
+	/* initialize to something sane */
+	sc_cgroup_fds fds = sc_udev_open_cgroup_v1(self->security_tag);
+	if (fds.cgroup_procs_fd < 0) {
+		die("cannot prepare cgroup v1 device hierarchy");
+	}
+	self->v1.fds = fds;
+
+	/* Deny device access by default.
+	 *
+	 * Write 'a' to devices.deny to remove all existing devices that were added
+	 * in previous launcher invocations, then add the static and assigned
+	 * devices. This ensures that at application launch the cgroup only has
+	 * what is currently assigned. */
+	sc_dprintf(fds.devices_deny_fd, "a");
+}
+
+static void _sc_cgroup_v1_close(sc_device_cgroup * self)
+{
+	sc_cleanup_cgroup_fds(&self->v1.fds);
+}
+
+/**
+ * SC_MINOR_ANY is used to indicate any minor device.
+ */
+const int SC_MINOR_ANY = INT_MAX;
+
+static void _sc_cgroup_v1_allow(sc_device_cgroup * self, int kind, int major,
+				int minor)
+{
+	if (minor != SC_MINOR_ANY) {
+		sc_dprintf(self->v1.fds.devices_allow_fd,
```

Comment:
I assume sc_dprintf will die() if it does not work, right? i.e. no error checking here is ok

Reply:
Yes, it dies if fewer characters were written than expected.